### PR TITLE
Use more specific exceptions where appropriate

### PR DESCRIPTION
**Pull request by phenaproxima**
Addresses #61.

### CODE REVIEW

**Comment by mbaynton** [comment]
My thoughts on this one were requested. Sorry my attention was pre-empted while discussing on the call, but for me it boils down to

- Throwing `TufException`s as much as possible / essentially always would be a nice goal for the advantage that users of the library would be able to easily isolate the cause of failed updates to php-tuf.
- That said, I don't personally mind breaking with that goal and throwing normal predefined exception types such as `\RuntimeException` for occurrences that arise from programming errors and not from unexpected or otherwise problematic inputs. That looks like all these particular changes do, so I'm +1 on them.
- Other exception types, in particular from Guzzle, were brought up as a reason we wouldn't be able to constrain our exceptions due to unexpected or otherwise problematic inputs to `TufException`s. If we've gone ahead with adding Guzzle as a dependency, my 2 cents is it'd be cool if we could cover common exceptions Guzzle might throw and rethrow those as TufExceptions, as this might help consumers of our library to integrate it successfully.

File: src/KeyDB.php [hunk 1]
```diff
@@ -151,15 +152,15 @@ public function addKey(array $keyMeta)
...
+     * @throws \InvalidArgumentException
      *     Thrown if the key ID is not found in the keydb database.
      *
      * @see https://github.com/theupdateframework/specification/blob/master/tuf-spec.md#4-document-formats
      */
     public function getKey(string $keyId)
     {
         if (empty($this->keys[$keyId])) {
-            throw new \Exception("Unknown key ID: $keyId");
+            throw new \InvalidArgumentException("Unknown key ID: $keyId");
```
Comment (mbaynton) [inline src/KeyDB.php:163]: Hmm, actually I'm not sure about this one. Key IDs are often read from other TUF repository metadata, so I sort of expect that a malformed remote TUF repository that referenced a nonexistent key would end up throwing here. To my mind that's in the class of bad input problems, not programming problems.

Comment (mbaynton) [reply src/KeyDB.php:163]: Likewise with the unknown role name. What about a `TufException` subclass for all malformed TUF repository issues?

Comment (phenaproxima) [reply src/KeyDB.php]: I ended up using the pre-existing `\Tuf\Exception\NotFoundException`, which seems to be the right fit for these errors.

**Review by mbaynton — APPROVED** [review]

File: src/JsonNormalizer.php [hunk 2]
```diff
@@ -33,15 +33,15 @@ public static function asNormalizedJson(array $structure) : string
...
+     * @throws \RuntimeException
      *     Thrown if sorting the array fails.
      *
      * @return void
      */
     private static function rKeySort(array &$structure) : void
     {
         if (!ksort($structure, SORT_STRING)) {
-            throw new \Exception("Failure sorting keys. Canonicalization is not possible.");
+            throw new \RuntimeException("Failure sorting keys. Canonicalization is not possible.");
```
Comment (xjm) [inline src/JsonNormalizer.php:44]: So this lightweight JSON canonicalization that we're implementing is a source of problems (exactly related to what David and Ted were debugging today), and this seems like a way to insert a vulnerability. "Feed it garbage JSON that causes a sort error, and then halt their updater." I feel we might want to step up a level at this whole API is a separate issue.

File: src/RoleDB.php [hunk 3]
```diff
@@ -61,67 +62,12 @@ public static function createFromRootMetadata(RootMetadata $rootMetadata)
...
-     *     The role name.
-     *
-     * @return boolean
-     *     True if the role is found in the role database; false otherwise.
-     */
-    public function roleExists(string $roleName)
-    {
-        return !empty($this->roles[$roleName]);
-    }
-
```
Comment (xjm) [inline src/RoleDB.php:124]: I'd rather remove this any other such dead code in a separate, dedicated issue, with code commends and followup issues to revisit it (similar to https://github.com/php-tuf/php-tuf/issues/21 ).

File: src/RoleDB.php [hunk 4]
```diff
@@ -163,25 +109,4 @@ public function getRoleKeyIds(string $roleName)
...
-     *     Thrown if the role does not exist.
-     */
-    public function getRoleThreshold(string $roleName)
-    {
-        if (! $this->roleExists($roleName)) {
-            throw new \Exception("Role does not exist: $roleName");
-        }
-
-        return $this->roles[$roleName]['threshold'];
-    }
```
Comment (xjm) [inline src/RoleDB.php:186]: More mostly dead code (to discuss separately as above).

**Review by xjm — COMMENTED** [review]
The rest looks pretty good; I like where this is going.

**Review by tedbow — CHANGES_REQUESTED** [review]
This needs to merged with main. Also I think the remove of code should be in separate issue. I think the removal would cause failures

File: src/RoleDB.php (hunk 3 above)
Comment (phenaproxima) [reply src/RoleDB.php]: Restored all this.

File: src/RoleDB.php (hunk 4 above)
Comment (phenaproxima) [reply src/RoleDB.php]: Restored for now.

**Comment by phenaproxima** [comment]
OK, I merged `main` back into this branch and restored (for now) the dead code, for removal in a different issue.

**Comment by xjm** [comment]
Can we file said different issue for dead code removal before merging this?

**Comment by phenaproxima** [comment]
Filed #102 for that.

File: src/RoleDB.php [hunk 5]
```diff
@@ -16,7 +17,7 @@ class RoleDB
      *
      * @var \array[]
      */
-    protected $roles;
+    protected $roles = [];
```
Comment (tedbow) [inline src/RoleDB.php:20]: Unrelated change

File: src/RoleDB.php [hunk 6]
```diff
@@ -61,7 +62,7 @@ public static function createFromRootMetadata(RootMetadata $rootMetadata)
             }
             $roleInfo['paths'] = [];
 
-            $db->addRole($roleName, $roleInfo);
+            $db->roles[$roleName] = $roleInfo;
```
Comment (tedbow) [inline src/RoleDB.php:65]: Unrelated change

**Review by tedbow — CHANGES_REQUESTED** [review]

**Review by tedbow — APPROVED** [review]
Looks good
